tests/chainer_tests/functions_tests/pooling_tests/test_average_pooling_nd.py

```diff
@@ -0,0 +1,180 @@
+    return [tuple(slice(max(x, 0), min(x + k, d))
+                  for (x, d, k) in six.moves.zip(xs, dims, ksize))
+            for xs in xss]
+
+
+@testing.parameterize(*testing.product({
+    'dims': [(4,), (4, 3), (4, 3, 2)],
+    'dtype': [numpy.float16, numpy.float32, numpy.float64],
+}))
+class TestAveragePoolingND(unittest.TestCase):
+
+    def setUp(self):
+        self.ndim = len(self.dims)
+        self.ksize = (3,) * self.ndim
+        self.stride = (2,) * self.ndim
+        self.pad = (1,) * self.ndim
+
+        x_shape = (2, 3) + self.dims
+        self.x = numpy.random.uniform(-1, 1, x_shape).astype(self.dtype)
+
+        outs = tuple(conv.get_conv_outsize(d, k, s, p, False)
+                     for (d, k, s, p) in six.moves.zip(
+                         self.dims, self.ksize, self.stride, self.pad))
+        gy_shape = (2, 3) + outs
+        self.gy = numpy.random.uniform(-1, 1, gy_shape).astype(self.dtype)
+
+        self.check_forward_options = {}
+        self.check_backward_options = {'eps': 1e-2}
+        if self.dtype == numpy.float16:
+            self.check_forward_options = {'atol': 5e-4, 'rtol': 5e-3}
+            self.check_backward_options = {
+                'eps': 1e-1, 'atol': 5e-3, 'rtol': 5e-2}
+
+    def check_forward(self, x_data, use_cudnn=True):
+        dims = self.dims
+        ksize = self.ksize
+        stride = self.stride
+        pad = self.pad
+        x = chainer.Variable(x_data)
+        y = functions.average_pooling_nd(x, ksize, stride, pad,
+                                         use_cudnn=use_cudnn)
+        self.assertEqual(y.data.dtype, self.dtype)
+        y_data = cuda.to_cpu(y.data)
+
+        self.assertEqual(self.gy.shape, y_data.shape)
+        for k in six.moves.range(2):
+            for c in six.moves.range(3):
+                x = self.x[k, c]
+                size = functools.reduce(operator.mul, ksize)
+                expect = numpy.array([x[idx].sum() for idx in pooling_patches(
+                    dims, ksize, stride, pad)])
+                expect = expect.reshape(y_data.shape[2:]) / size
+                testing.assert_allclose(
+                    expect, y_data[k, c], **self.check_forward_options)
+
+    @condition.retry(3)
+    def test_forward_cpu(self):
+        self.check_forward(self.x)
+
+    @attr.cudnn
+    @condition.retry(3)
+    def test_forward_gpu(self):
+        self.check_forward(cuda.to_gpu(self.x))
+
+    @attr.gpu
+    @condition.retry(3)
+    def test_forward_gpu_no_cudnn(self):
+        self.check_forward(cuda.to_gpu(self.x), False)
+
+    def test_forward_consistency_regression(self):
+        # Regression test to average_pooling_2d.
+
+        if len(self.dims) != 2:
+            return
+
+        ksize = self.ksize
+        stride = self.stride
+        pad = self.pad
+
+        y_nd = functions.average_pooling_nd(self.x, ksize, stride=stride,
+                                            pad=pad, use_cudnn=False)
```

Comment:
I've add the test with `use_cudnn=True` as well as the test with CuPy condition.